feat(api): Heater Shaker Firmware Level Emulator

DerekMaggio opened the pull request:
# Overview

Add firmware level emulator and tests for Heater Shaker

# Changelog

- Create HeaterShakerEmulator class
- Create RPM simulation
- Add Heater/Shaker settings
- Add Heater/Shaker to monorepo docker compose file

# Review requests

- @sanni-t & @pmoegenburg can you do a deep dive into whether or not my emulation accurately reproduces the functionality of the hardware Heater/Shaker
- This is my first PR since being gone on paternity leave, I probably made some stupid errors and style mistakes


# Risk assessment

Low

Commit:
feat: Heater-Shaker Firmware Emulator




remove files


Fix some parsing errors


Formatting


Change rpm_per_tick

Commit:
Tests

DerekMaggio commented:
I confirmed that what I run the docker-compose.yaml file in the monorepo everything starts correctly and I am able to hit the /modules endpoint successfully.

sanni-t reviewed (commented):
This is great! Tested the docker-built emulator with dev server and it works! (simply running emulator app with `make -C api emulator` results in an unrelated error but I think it's been like that for a while now)

Two things I noticed- 
1. Sending a `deactivate_shaker` makes the heater-shaker in emulator unresponsive for a few seconds. It eventually responds with shake deactivated though so was wondering if that's an intentional silence. 
2. `deactivate_heater` is not yet implemented. Left out intentionally? 

Thanks so much for implementing this so quickly!

Commit:
Modify home functionality


Formatting and linting

DerekMaggio commented:
So after much fruitless investigation, I have no idea why the emulator does not handle the `deactivate_shaker` command from the [Heater/Shaker Hardware Controller](https://github.com/Opentrons/opentrons/blob/edge/api/src/opentrons/hardware_control/modules/heater_shaker.py#L428-L432).

Had a meeting with Robot Services and they are going to take over the task of getting the command to work.

DerekMaggio commented:
@sanni-t please note that I have changed the base from `edge` to `heater-shaker-firmware-emulator-wip`.
I don't want to merge this to edge until it is completely working

jbleon95 commented on api/src/opentrons/hardware_control/emulation/heater_shaker.py:
```diff
@@ -0,0 +1,138 @@
+        return "M3"
+
+    def _get_rpm(self, command: Command) -> str:
+        res = f"M123 C:{self._rpm.current} T:{self._rpm.target}"
```
This line is causing the tests to hang, since the `parse_rpm_response` util function is only set up to handle an int (even though confusingly it sets the target type as `Optional[int]`). Without making changes to that function to properly handle a `None`/`'none'` value, this can pass if we make sure we are returning a `0` if `self._rpm.target` is `None`

SyntaxColoring commented on api/tests/opentrons/hardware_control/integration/test_heatershaker.py:
```diff
@@ -0,0 +1,72 @@
+
+async def test_latch_status(heatershaker: HeaterShaker) -> None:
+    """It should run open and close latch."""
+    await heatershaker.wait_next_poll()
```
Why do we need to do `wait_next_poll()` at the beginning of all of these tests?

DerekMaggio replied:
To be honest, I am not sure. 
I am pretty unfamiliar with the async stuff so I copied and pasted the tests from test_thermocycley.py and modified them.

DerekMaggio replied:
My reasoning that it is there is to make sure that the emulator inside of conftest was in fact started.

DerekMaggio replied:
If I remove the calls the tests fail because the emulator does not initialize correctly

SyntaxColoring replied:
What's the nature of that failure? Is it because `heatershaker.labware_latch_status.value` would be `None` or unknown or something, and `wait_next_poll()` makes sure it's populated with an actual value?

DerekMaggio replied:
Correct

jbleon95 committed:
fix rpm emulator g-code and add deactivate test

Commit:
Fix tests

DerekMaggio commented:
@sanni-t can I get a re-review?

Commit:
Add deactivate_heater G-Code

SyntaxColoring commented on api/tests/opentrons/hardware_control/integration/test_heatershaker.py:
```diff
@@ -0,0 +1,72 @@
+    await heatershaker.wait_next_poll()
+    await heatershaker.start_set_temperature(50.0)
+
+    # Have to wait for next poll because target temp will not update until then
```
Is this the intended API contract of the Heater-Shaker hardware controller? If I do `hs.start_set_temperature(...)` and then do `t = hs.target_temperature`, I'm not guaranteed to get back what I just set?

If yes: I think this is surprising, and we need to document it in the `HeaterShaker` class. Doesn't have to happen in this PR, but I'd like confirmation/consensus.

If no: this indicates a bug, I guess?

DerekMaggio replied:
You know, it makes sense that target temp should be set and it would be just current temp that hasn't update yet.

Looking into it now.

DerekMaggio replied:
I left out a line. Fixing it in a followup commit

DerekMaggio replied:
https://github.com/Opentrons/opentrons/pull/11121/commits/4beaaca56907c83a335fe01afdee6903f3174e65

Commit:
Add start_set_temperature

sanni-t commented:
The behavior is almost there!
Two things that need to be cleaned up:
1. On the real hardware, a home command (G28) will only respond back once homing is finished. If we try to do that with the ramp-down behavior of deactivate then it'll take quite a long time, unlike the real hardware that goes from shaking to homed in just 2-3 seconds. So I think it'll be okay to not use the speed ramp down for homing. Instead we can add a delay of 2-3 seconds, set the target and current speeds to zero, and respond with `G28 OK`.
2. Ramping down of current temperature on deactivating heater (you probably already know). This one will respond to the gcode immediately after deactivation and then continue ramping down to an ambient temperature.

Commit:
Update home and deactivate_heater functionality

Based off of https://github.com/Opentrons/opentrons/pull/11121#issuecomment-1190383124

DerekMaggio commented:
> The behavior is almost there! Two things that need to be cleaned up:
> 
> 1. On the real hardware, a home command (G28) will only respond back once homing is finished. If we try to do that with the ramp-down behavior of deactivate then it'll take quite a long time, unlike the real hardware that goes from shaking to homed in just 2-3 seconds. So I think it'll be okay to not use the speed ramp down for homing. Instead we can add a delay of 2-3 seconds, set the target and current speeds to zero, and respond with `G28 OK`.
> 2. Ramping down of current temperature on deactivating heater (you probably already know). This one will respond to the gcode immediately after deactivation and then continue ramping down to an ambient temperature.

https://github.com/Opentrons/opentrons/pull/11121/commits/26f98cdbc0ab76c8a8938feae5977f0ccafb0ad9

DerekMaggio commented:
@sanni-t I addressed your comments [here](https://github.com/Opentrons/opentrons/pull/11121/commits/26f98cdbc0ab76c8a8938feae5977f0ccafb0ad9)


Is it okay to use a regular sleep in the _home function of the emulator?

Commit:
Linting

sanni-t commented:
> @sanni-t I addressed your comments [here](https://github.com/Opentrons/opentrons/pull/11121/commits/26f98cdbc0ab76c8a8938feae5977f0ccafb0ad9)
> 
🙌

> Is it okay to use a regular sleep in the _home function of the emulator?
That is a good point. It won't matter for manual testing but it would for the unit tests & automated tests. What do we do for unit tests that need to wait for, say, speed to ramp down? Are we changing the duration of a tick to something very small? In that case, maybe we could modify the tick function to ramp down with `rpm_per_tick` of something large, like 500.. only when target is 0.. Just a thought

sanni-t commented on api/src/opentrons/hardware_control/emulation/heater_shaker.py:
```diff
@@ -0,0 +1,146 @@
+        return f"M241 STATUS:{self._latch_status.value.upper()}"
+
+    def _deactivate_heater(self, command: Command) -> str:
+        self._temperature.set_target(TEMPERATURE_ROOM)
```
This makes the temperature status respond with a target of room temperature value. We want the target to return as `None`. It looks like the thermocycler & tempdeck's deactivate emulator simply sets the target to `None` and current temp to room temperature. So, I'm fine with implementing the same thing here for now. We can revisit them all later.

DerekMaggio commented:
Closing to gather requirements and make sure functionality is correct

sfoster1 commented:
> > That is a good point. It won't matter for manual testing but it would for the unit tests & automated tests. What do we do for unit tests that need to wait for, say, speed to ramp up? Are we changing the duration of a tick to something very small? In that case, maybe we could modify the tick function to ramp down with `rpm_per_tick` of something large, like 500.. only when target is 0.. Just a thought
> 
> Or if that isn't easy to implement, just set the target & current rpm to 0 and send the gcode response. And leave a TODO note to improve this behavior to mimic the hardware's true behavior. We can address all improvements in a follow-up ticket when we're not pressed for time.

Let's do the latter. We really need to resist the urge to implement too much behavior here. This is a testing aid and nothing more, and even then because this interacts with a physical system and we're not simulating the world, there's always going to be little places where the behavior doesn't match. We should definitely make the exact format of communications on the wire the same as firmware, like changing an unset target to `None`; we should definitely not stray into little bits of physical modeling.

DerekMaggio commented:
There seems to be something going on between the temperature/RPM backends that the emulator is using and the poller the module is using to query the backend. 

Specifically around deactivating the heater and the shaker. 
I deactivate it but the poller does not update with the correct values.

For instance, when deactivating the heater I am setting the temperature backend to the following: Target: None, Current: Room Temperature. 
But the poller that the module uses to get it's values never updates to the correct values.

Can anyone help me out with this?

DerekMaggio commented:
> There seems to be something going on between the temperature/RPM backends that the emulator is using and the poller the module is using to query the backend.
> 
> Specifically around deactivating the heater and the shaker. I deactivate it but the poller does not update with the correct values.
> 
> For instance, when deactivating the heater I am setting the temperature backend to the following: Target: None, Current: Room Temperature. But the poller that the module uses to get it's values never updates to the correct values.
> 
> Can anyone help me out with this?

Never mind, I was setting rpm to room temp, not temp to room temp.

Commit:
fix: Fix deactivate_heater function

Commit:
Make home_delay_time customizable

DerekMaggio commented:
@sanni-t 
I made it so when you call deactivate_heater it now sets the current temp to TEMPERATURE_ROOM and sets target to None.
I also make the home delay time customizable by adding a home_delay_time to the HeaterShakerSettings object

Commit:
Linting

sanni-t reviewed (approved):
This works as expected! Thank you!
